File: internal/api/jdwp/stringreference.go
package jdwp

import (
	"fmt"

	"cli-debugger/pkg/errors"
)

// StringReference Command Set Implementation
// StringReference command constants are already defined in protocol.go

// Value Get string value
func (c *Client) Value(stringID string) (string, error) {
	data := encodeID(stringID, c.idsizes.ObjectIDSize)

	packet := createCommandPacketWithData(stringReferenceCommandSet, stringReferenceCommandValue, data)
	if err := c.sendPacket(packet); err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get string value")
	}

	reply, err := c.readReply()
	if err != nil {
		return "", errors.WrapCommandError(err, errors.ErrCommandFailed, "Failed to get string value")
	}

	if reply.ErrorCode != 0 {
		return "", errors.NewProtocolError(errors.ErrProtocolError,
			fmt.Sprintf("Get string value failed: %s", reply.Message))
	}

	reader := newPacketReader(reply.Data)
	value, err := reader.readString()
	if err != nil {
		return "", errors.WrapProtocolError(err, errors.ErrInvalidResponse, "Failed to read string value")
	}

	return value, nil
}

File: src/protocol/jdwp/string-reference.ts
/**
 * StringReference Command Set Implementation
 */

import {
  CommandSet,
  StringReferenceCommand,
  type IDSizes,
} from "./protocol.js";
import { createCommandPacketWithData, encodeID } from "./codec.js";
import { PacketReader } from "./reader.js";
import { APIError, ErrorType, ErrorCodes } from "../errors.js";

/**
 * JDWP Client interface for command execution
 */
export interface JDWPCommandExecutor {
  sendPacket(packet: Buffer): Promise<void>;
  readReply(): Promise<{ errorCode: number; message: string; data: Buffer }>;
  idSizes: IDSizes;
}

/**
 * Get string value
 */
export async function getStringValue(
  executor: JDWPCommandExecutor,
  stringID: string,
): Promise<string> {
  const data = encodeID(stringID, executor.idSizes.objectIDSize);
  const packet = createCommandPacketWithData(
    CommandSet.StringReference,
    StringReferenceCommand.Value,
    data,
  );
  await executor.sendPacket(packet);

  const reply = await executor.readReply();
  if (reply.errorCode !== 0) {
    throw new APIError(
      ErrorType.ProtocolError,
      ErrorCodes.ProtocolError,
      `Get string value failed: ${reply.message}`,
    );
  }

  const reader = new PacketReader(reply.data);
  return reader.readString();
}
